tests/rest/admin/test_room.py

```diff
@@ -1419,31 +1421,30 @@ def _order_test(
             "joined_local_members", [room_id_1, room_id_2, room_id_3], reverse=True
         )
 
-        _order_test("version", all_ids)
-        _order_test("version", all_ids, reverse=True)
+        _order_test("version", sorted_by_room_id_asc, reverse=True)
+        _order_test("version", sorted_by_room_id_desc)
```

Comment:
I have no idea why this requires the reverse to be swapped. It was required locally, maybe CI tests differently, but I cannot explain it since all rooms should be the same version, and reversing works the same (and correctly) for the rest of the test cases

Reply:
That's odd. Thanks for flagging it up; let's see what CI says and maybe we can investigate in more detail later.

Reply:
Ahh, I think I see it now. The initial sort order is controlled by `order_by_asc`, which differs between different sorting orders. See e.g.:

https://github.com/matrix-org/synapse/blob/d2e595f379720ac0fc8225db1cd999fe31e76cfa/synapse/storage/databases/main/room.py#L506-L511

`dir=b` reverses this rather than setting it to False.

I guess the idea is to use the "natural" sort direction by default while still allowing the user to reverse it. Not sure if I agree with that choice, but it looks like the behaviour will be consistent.

Reply:
(This goes back to #6720)

Reply:
> I guess the idea is to use the "natural" sort direction by default while still allowing the user to reverse it. 

That indeed was the intention, and back then there were only two sort orders - size of joined members and alphabetical room names. It seemed sensible for users to want to view rooms sorted by size of joined members descending, whereas users would want to see room names sorted from a->z.

Of course, client software could just set the `dir=b` flag itself when sorting alphabetically, but that's extra work for the client. Default directions are noted in the [docs](https://matrix-org.github.io/synapse/latest/admin_api/rooms.html#list-room-api).

I'm just coming into this conversation without much context... but I fail to see why setting the direction order based on ordering type is confusing.

Reply:
> but I fail to see why setting the direction order based on ordering type is confusing.

In a nutshell, we have different interpretations of what `dir=b` ought to mean
- as-implemented: reverse the natural ordering
- our guess: sort the results descending by the given criterion

Reply:
Aah I see, yes, in my case this will end up confusing:

![image](https://user-images.githubusercontent.com/1695179/149565447-349e5154-fc28-4ae7-975c-aa3833e751ef.png)

Because the arrow directly translates to `b` or `f` (as usually done through APIs, but usually `asc` `desc` where meaning is well defined), and there is some meaning behind the arrow. So now, blindly translating the direction to `b` or `f` will give the wrong arrow direction. This could be fixed client-side by keeping track of the natural way of sorting, but then the logic becomes more complex and based on how it's written in Synapse. (One could argue it's already tightly bound to synapse as it works with this specific API, but it makes the sorting non-trivial)

Reply:
Personally, I feel like the initial sort direction is the choice of the client, and the backend just returns exactly what you ask it. That way, the client can be written knowing exactly what it will return without looking at the docs/source code and without having some lookup table to correctly swap arrows based on field. Because at that stage, you're essentially doing double work: backend is opinionated about sorting direction, so will pick a (well defined, but seemingly arbitrary) direction that can be flipped, and client compensates for this opinion by swapping the arrow to match the direction.

Reply:
Thanks for the discussion both - I've ended up coming to the same conclusion. It is unnecessary to maintain this "lookup table" on both ends.

I've created a separate issue to track this at https://github.com/matrix-org/synapse/issues/11759.